File: samples/conference/registration-bc/src/main/java/com/microsoft/conference/registration/readmodel/OrderViewModelGenerator.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using System.Threading.Tasks;
using Conference.Common;
using ECommon.Components;
using ECommon.Dapper;
using ECommon.IO;
using ENode.Infrastructure;
using Registration.Orders;

namespace Registration.Handlers
{
    [Component]
    public class OrderViewModelGenerator :
        IMessageHandler<OrderPlaced>,
        IMessageHandler<OrderRegistrantAssigned>,
        IMessageHandler<OrderReservationConfirmed>,
        IMessageHandler<OrderPaymentConfirmed>,
        IMessageHandler<OrderExpired>,
        IMessageHandler<OrderClosed>,
        IMessageHandler<OrderSuccessed>
    {
        public Task<AsyncTaskResult> HandleAsync(OrderPlaced evnt)
        {
            return TryTransactionAsync((connection, transaction) =>
            {
                var tasks = new List<Task>();

                //插入订单主记录
                tasks.Add(connection.InsertAsync(new
                {
                    OrderId = evnt.AggregateRootId,
                    ConferenceId = evnt.ConferenceId,
                    Status = (int)OrderStatus.Placed,
                    AccessCode = evnt.AccessCode,
                    ReservationExpirationDate = evnt.ReservationExpirationDate,
                    TotalAmount = evnt.OrderTotal.Total,
                    Version = evnt.Version
                }, ConfigSettings.OrderTable, transaction));

                //插入订单明细
                foreach (var line in evnt.OrderTotal.Lines)
                {
                    tasks.Add(connection.InsertAsync(new
                    {
                        OrderId = evnt.AggregateRootId,
                        SeatTypeId = line.SeatQuantity.Seat.SeatTypeId,
                        SeatTypeName = line.SeatQuantity.Seat.SeatTypeName,
                        Quantity = line.SeatQuantity.Quantity,
                        UnitPrice = line.SeatQuantity.Seat.UnitPrice,
                        LineTotal = line.LineTotal
                    }, ConfigSettings.OrderLineTable, transaction));
                }

                return tasks;
            });
        }
        public Task<AsyncTaskResult> HandleAsync(OrderRegistrantAssigned evnt)
        {
            return TryUpdateRecordAsync(connection =>
            {
                return connection.UpdateAsync(new
                {
                    RegistrantFirstName = evnt.Registrant.FirstName,
                    RegistrantLastName = evnt.Registrant.LastName,
                    RegistrantEmail = evnt.Registrant.Email,
                    Version = evnt.Version
                }, new
                {
                    OrderId = evnt.AggregateRootId,
                    Version = evnt.Version - 1
                }, ConfigSettings.OrderTable);
            });
        }
        public Task<AsyncTaskResult> HandleAsync(OrderReservationConfirmed evnt)
        {
            return TryUpdateRecordAsync(connection =>
            {
                return connection.UpdateAsync(new
                {
                    Status = (int)evnt.OrderStatus,
                    Version = evnt.Version
                }, new
                {
                    OrderId = evnt.AggregateRootId,
                    Version = evnt.Version - 1
                }, ConfigSettings.OrderTable);
            });
        }
        public Task<AsyncTaskResult> HandleAsync(OrderPaymentConfirmed evnt)
        {
            return TryUpdateRecordAsync(connection =>
            {
                return connection.UpdateAsync(new
                {
                    Status = (int)evnt.OrderStatus,
                    Version = evnt.Version
                }, new
                {
                    OrderId = evnt.AggregateRootId,
                    Version = evnt.Version - 1
                }, ConfigSettings.OrderTable);
            });
        }
        public Task<AsyncTaskResult> HandleAsync(OrderExpired evnt)
        {
            return TryUpdateRecordAsync(connection =>
            {
                return connection.UpdateAsync(new
                {
                    Status = (int)OrderStatus.Expired,
                    Version = evnt.Version
                }, new
                {
                    OrderId = evnt.AggregateRootId,
                    Version = evnt.Version - 1
                }, ConfigSettings.OrderTable);
            });
        }
        public Task<AsyncTaskResult> HandleAsync(OrderClosed evnt)
        {
            return TryUpdateRecordAsync(connection =>
            {
                return connection.UpdateAsync(new
                {
                    Status = (int)OrderStatus.Closed,
                    Version = evnt.Version
                }, new
                {
                    OrderId = evnt.AggregateRootId,
                    Version = evnt.Version - 1
                }, ConfigSettings.OrderTable);
            });
        }
        public Task<AsyncTaskResult> HandleAsync(OrderSuccessed evnt)
        {
            return TryUpdateRecordAsync(connection =>
            {
                return connection.UpdateAsync(new
                {
                    Status = (int)OrderStatus.Success,
                    Version = evnt.Version
                }, new
                {
                    OrderId = evnt.AggregateRootId,
                    Version = evnt.Version - 1
                }, ConfigSettings.OrderTable);
            });
        }

        private async Task<AsyncTaskResult> TryUpdateRecordAsync(Func<IDbConnection, Task<int>> action)
        {
            using (var connection = GetConnection())
            {
                await action(connection);
                return AsyncTaskResult.Success;
            }
        }
        private async Task<AsyncTaskResult> TryTransactionAsync(Func<IDbConnection, IDbTransaction, IEnumerable<Task>> actions)
        {
            using (var connection = GetConnection())
            {
                await connection.OpenAsync().ConfigureAwait(false);
                var transaction = await Task.Run<SqlTransaction>(() => connection.BeginTransaction()).ConfigureAwait(false);
                try
                {
                    await Task.WhenAll(actions(connection, transaction)).ConfigureAwait(false);
                    await Task.Run(() => transaction.Commit()).ConfigureAwait(false);
                    return AsyncTaskResult.Success;
                }
                catch
                {
                    transaction.Rollback();
                    throw;
                }
            }
        }
        private SqlConnection GetConnection()
        {
            return new SqlConnection(ConfigSettings.ConferenceConnectionString);
        }
    }
}

File: samples/conference/registration-bc/src/main/java/com/microsoft/conference/registration/readmodel/OrderViewModelGenerator.java
package com.microsoft.conference.registration.readmodel;

import com.baomidou.mybatisplus.core.conditions.update.LambdaUpdateWrapper;
import com.microsoft.conference.common.dataobject.OrderDO;
import com.microsoft.conference.common.dataobject.OrderLineDO;
import com.microsoft.conference.common.mapper.OrderLineMapper;
import com.microsoft.conference.common.mapper.OrderMapper;
import com.microsoft.conference.registration.domain.order.event.OrderClosed;
import com.microsoft.conference.registration.domain.order.event.OrderExpired;
import com.microsoft.conference.registration.domain.order.event.OrderPaymentConfirmed;
import com.microsoft.conference.registration.domain.order.event.OrderPlaced;
import com.microsoft.conference.registration.domain.order.event.OrderRegistrantAssigned;
import com.microsoft.conference.registration.domain.order.event.OrderReservationConfirmed;
import com.microsoft.conference.registration.domain.order.event.OrderSuccessed;
import com.microsoft.conference.registration.domain.order.model.OrderStatus;
import org.enodeframework.annotation.Event;
import org.springframework.beans.factory.annotation.Autowired;

/**
 *
 */
@Event
public class OrderViewModelGenerator {

    @Autowired
    private OrderMapper orderMapper;

    @Autowired
    private OrderLineMapper orderLineMapper;

    public void handleAsync(OrderPlaced evnt) {

        //插入订单主记录
        OrderDO orderDO = OrderConvert.INSTANCE.toDO(evnt);
        orderDO.setStatus(OrderStatus.Placed.getStatus());
        orderMapper.insert(orderDO);

        //插入订单明细
        evnt.getOrderTotal().getOrderLines().forEach(orderLine -> {
            OrderLineDO orderLineDO = OrderConvert.INSTANCE.toDO(evnt, orderLine);
            orderLineMapper.insert(orderLineDO);
        });

    }


    public void handleAsync(OrderRegistrantAssigned evnt) {
        OrderDO orderDO = OrderConvert.INSTANCE.toDO(evnt);
        LambdaUpdateWrapper<OrderDO> wrapper = new LambdaUpdateWrapper<>();
        wrapper.eq(OrderDO::getOrderId, evnt.getAggregateRootId());
        wrapper.eq(OrderDO::getVersion, evnt.getVersion() - 1);
        orderMapper.update(orderDO, wrapper);
    }

    public void handleAsync(OrderReservationConfirmed evnt) {
        OrderDO orderDO = new OrderDO();
        orderDO.setStatus(evnt.getOrderStatus().getStatus());
        orderDO.setVersion(evnt.getVersion());
        LambdaUpdateWrapper<OrderDO> wrapper = new LambdaUpdateWrapper<>();
        wrapper.eq(OrderDO::getOrderId, evnt.getAggregateRootId());
        wrapper.eq(OrderDO::getVersion, evnt.getVersion() - 1);
        orderMapper.update(orderDO, wrapper);
    }

    public void handleAsync(OrderPaymentConfirmed evnt) {
        OrderDO orderDO = new OrderDO();
        orderDO.setStatus(evnt.getOrderStatus().getStatus());
        orderDO.setVersion(evnt.getVersion());
        LambdaUpdateWrapper<OrderDO> wrapper = new LambdaUpdateWrapper<>();
        wrapper.eq(OrderDO::getOrderId, evnt.getAggregateRootId());
        wrapper.eq(OrderDO::getVersion, evnt.getVersion() - 1);
        orderMapper.update(orderDO, wrapper);
    }

    public void handleAsync(OrderExpired evnt) {
        OrderDO orderDO = new OrderDO();
        orderDO.setStatus(OrderStatus.Expired.getStatus());
        orderDO.setVersion(evnt.getVersion());
        LambdaUpdateWrapper<OrderDO> wrapper = new LambdaUpdateWrapper<>();
        wrapper.eq(OrderDO::getOrderId, evnt.getAggregateRootId());
        wrapper.eq(OrderDO::getVersion, evnt.getVersion() - 1);
        orderMapper.update(orderDO, wrapper);
    }

    public void handleAsync(OrderClosed evnt) {
        OrderDO orderDO = new OrderDO();
        orderDO.setStatus(OrderStatus.Closed.getStatus());
        orderDO.setVersion(evnt.getVersion());
        LambdaUpdateWrapper<OrderDO> wrapper = new LambdaUpdateWrapper<>();
        wrapper.eq(OrderDO::getOrderId, evnt.getAggregateRootId());
        wrapper.eq(OrderDO::getVersion, evnt.getVersion() - 1);
        orderMapper.update(orderDO, wrapper);
    }

    public void handleAsync(OrderSuccessed evnt) {
        OrderDO orderDO = new OrderDO();
        orderDO.setStatus(OrderStatus.Success.getStatus());
        orderDO.setVersion(evnt.getVersion());
        LambdaUpdateWrapper<OrderDO> wrapper = new LambdaUpdateWrapper<>();
        wrapper.eq(OrderDO::getOrderId, evnt.getAggregateRootId());
        wrapper.eq(OrderDO::getVersion, evnt.getVersion() - 1);
        orderMapper.update(orderDO, wrapper);
    }
}
